src/main/java/org/jenkinsci/plugins/vsphere/FixedLifespanCloudRetentionStrategy.java

```diff
@@ -0,0 +1,114 @@
+package org.jenkinsci.plugins.vsphere;
+
+import hudson.model.Descriptor;
+import hudson.model.InvisibleAction;
+import hudson.slaves.AbstractCloudComputer;
+import hudson.slaves.AbstractCloudSlave;
+import hudson.slaves.CloudRetentionStrategy;
+import hudson.slaves.RetentionStrategy;
+import hudson.util.TimeUnit2;
+
+import java.io.IOException;
+import java.util.concurrent.ExecutionException;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+
+import org.kohsuke.accmod.Restricted;
+import org.kohsuke.accmod.restrictions.NoExternalUse;
+import org.kohsuke.stapler.DataBoundConstructor;
+
+import static java.util.logging.Level.WARNING;
+
+public final class FixedLifespanCloudRetentionStrategy extends RetentionStrategy<AbstractCloudComputer> {
+
+    private static final Logger LOGGER = Logger.getLogger(CloudRetentionStrategy.class.getName());
+
+    private final int lifespanMinutes;
+
+    @DataBoundConstructor
+    public FixedLifespanCloudRetentionStrategy(int lifespanMinutes) {
+        this.lifespanMinutes = lifespanMinutes;
+    }
+
+    public int getLifespanMinutes() {
+        return lifespanMinutes;
+    }
+
+    @Override
+    public long check(AbstractCloudComputer c) {
+        final long creationTimeMillis = c.getAction(CloudComputerCreatedOnInvisibleAction.class).getCreationTimeMillis();
+        final long ageMillis = System.currentTimeMillis() - creationTimeMillis;
+        final String cname = c.getName();
+        if (!isAtEndOfLife() && ageMillis > TimeUnit2.MINUTES.toMillis(lifespanMinutes)) {
+            LOGGER.log(Level.FINE, "Will terminate {0} once idle - lifespan of {1} minutes reached.", new Object[] { cname, lifespanMinutes });
+            setAtEndOfLife();
+            final VSphereOfflineCause cause = new VSphereOfflineCause(Messages._fixedLifespanCloudRetentionStrategy_OfflineReason_LifespanReached(String.valueOf(lifespanMinutes)));
+            try {
+                c.disconnect(cause).get();
+            } catch (InterruptedException | ExecutionException e) {
+                LOGGER.log(WARNING, "Failed to disconnect " + cname, e);
+            }
+        }
+        if (isAtEndOfLife() && c.isIdle()) {
+            final AbstractCloudSlave computerNode = c.getNode();
+            if (computerNode != null) {
+                try {
+                    LOGGER.log(Level.FINER, "Initiating termination of {0}.", cname);
+                    computerNode.terminate();
+                } catch (InterruptedException | IOException e) {
+                    LOGGER.log(WARNING, "Failed to terminate " + cname, e);
+                }
+            }
+        }
+        return 1; // re-check in 1 minute
```

Comment:
Once a machine has reached end-of-life then re-checking every minute is good.
However, before it's reached end-of-life, we don't need to check every minute - we should return "number of minutes until we reach end of line".